File: medium/cpp/776.cpp
// https://leetcode.com/problems/split-bst/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    std::vector<TreeNode*> splitBST(TreeNode* root, int target)
    {
        return solve(root, target);
    }

private:
    std::vector<TreeNode*> solve(TreeNode* root, int target)
    {
        if (root == nullptr)
        {
            return { nullptr, nullptr };
        }

        if (root->val <= target)
        {
            auto res = solve(root->right, target);
            root->right = res[0];
            return { root, res[1] };
        }
        else
        {
            auto res = solve(root->left, target);
            root->left = res[1];
            return { res[0], root };
        }
    }
};

File: medium/python/776.py
# https://leetcode.com/problems/split-bst/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def splitBST(self, root: Optional[TreeNode], target: int) -> List[Optional[TreeNode]]:
        return self.split(root, target)

    def split(self, root, target):
        if not root:
            return [None, None]

        if root.val > target:
            left, right = self.split(root.left, target)
            root.left = right
            return [left, root]
        else:
            left, right = self.split(root.right, target)
            root.right = left
            return [root, right]